crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs

```diff
@@ -0,0 +1,707 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use indexmap::IndexSet;
+use ndarray::{azip, s, Array1, Array2, ArrayView2};
+use smallvec::{smallvec, SmallVec};
+
+use crate::synthesis::linear::utils::calc_inverse_matrix_inner;
+use qiskit_circuit::operations::{Param, StandardGate};
+use qiskit_circuit::Qubit;
+
+/// Symplectic matrix.
+/// Currently this class is internal to the greedy synthesis algorithm.
+struct SymplecticMatrix {
+    /// Number of qubits.
+    num_qubits: usize,
+    /// Matrix with dimensions (2 * num_qubits) x (2 * num_qubits).
+    smat: Array2<bool>,
+}
+
+/// Clifford.
+/// Currently this class is internal to the greedy synthesis algorithm
+/// and has a very different functionality from Qiskit's python-based
+/// Clifford class.
+struct Clifford {
+    /// Number of qubits.
+    num_qubits: usize,
+    /// Matrix with dimensions (2 * num_qubits) x (2 * num_qubits + 1).
```

Comment:
what's the reason to have the phases as the last line of the matrix and not as a separate vector (unlike the current Clifford class)? will it be useful later?

Reply:
Hmm, this is an interesting suggestion, which has its pros and cons. The current form is consistent with how we store a Clifford tableau in Python, but we can change it of course. It is also easier to perform row operations directly on the (2n) x (2n+1) matrix.